caddyconfig/httpcaddyfile/directives.go

```diff
@@ -398,6 +398,21 @@ func sortRoutes(routes []ConfigValue) {
 		if len(jPM) > 0 {
 			jPathLen = len(jPM[0])
 		}
+
+		// if both directives have no path matcher, use whichever one
+		// has any kind of matcher defined first.
+		if iPathLen == 0 && jPathLen == 0 {
+			return len(iRoute.MatcherSetsRaw) > len(jRoute.MatcherSetsRaw)
+		}
+
+		// sort the "root" directive in reverse because we want the
+		// most-specific path last such that the root variable
+		// overrides a less-specific one.
+		if iDir == "root" {
+			return iPathLen < jPathLen
+		}
```

Comment:
I think we have to be careful before we decide to do this... though I'm not opposed, the implications here for consistency and correctness run deeper. By similar logic, we should also sort `header` directives this way. And who knows which other ones. If we're going to treat one as a special case for a certain reason, we need to treat all of them where the same logic applies: any middleware directive that effectively overwrites any instances of the previous directive.

Reply:
We can easily do that with a lookup table beside the other directive order map, but the question obviously is then how do we want it to work for plugins, do we want to give them an API to plug into this as well?

Reply:
Let's not bite off more than we can chew here... and just keep it simple for now.

Remember this is just a convenience, there's always the `route` directive to do things manually.

Reply:
Sounds like `root`, `rewrite` and `header`, not sure if we need `handle` and `handle_path` as well :thinking:

Reply:
Alright so we talked offline, didn't need this bit ultimately... the mutual exclusivity (the `group` thing) should make this work correctly anyways, we just needed the fix for non-path matchers for this to do the right thing.